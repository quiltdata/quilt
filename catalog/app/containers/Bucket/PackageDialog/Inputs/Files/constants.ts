import cfg from 'constants/config'

export const MAX_UPLOAD_SIZE = 20 * 1000 * 1000 * 1000 // 20GB
// XXX: keep in sync w/ the backend
// NOTE: these limits are lower than the actual "hard" limits on the backend
export const MAX_S3_SIZE = cfg.chunkedChecksums
  ? 5 * 10 ** 12 // 5 TB
  : 50 * 10 ** 9 // 50 GB
export const MAX_FILE_COUNT = 1000
