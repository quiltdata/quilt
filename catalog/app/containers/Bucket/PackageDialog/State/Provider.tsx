import invariant from 'invariant'
import * as React from 'react'

import type { FilesState } from './files'
import type { PackageSrc } from './manifest'
import { State, useState } from './State'

interface ProviderProps {
  children: React.ReactNode
  src?: PackageSrc
  dst: { bucket: string; name?: string }
  open?: boolean | FilesState['value']['added']
}

export function Provider({ children, dst, src, open = false }: ProviderProps) {
  const contextValue = useState(dst, src, open)
  return <Context.Provider value={contextValue}>{children}</Context.Provider>
}

export const Context = React.createContext<State | null>(null)

export function useContext(): State {
  const context = React.useContext(Context)
  invariant(context, 'useContext must be used within PackageDialogProvider')
  return context
}
