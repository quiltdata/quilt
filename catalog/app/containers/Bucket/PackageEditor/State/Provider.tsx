import * as React from 'react'

import { L } from 'components/Form/Package/types'
import type * as Types from 'utils/types'
import type { Schema } from 'utils/workflows'

import type { Manifest } from '../../PackageDialog/Manifest'

import useSource, { Src } from './Source'
import useManifest from './Manifest'
import useBucket, { BucketContext } from './Bucket'
import useWorkflow, { WorkflowContext } from './Workflow'

export interface InputState {
  errors?: Error[] | typeof L
  value: string
}

export interface MetaState {
  value?: Types.JsonRecord
  schema?: Schema
}

interface MessageContext {
  state: InputState | typeof L
  actions: {
    onChange: (v: string) => void
  }
}

interface MetaContext {
  state: MetaState | typeof L
}

interface NameContext {
  state: InputState | typeof L
  actions: {
    onChange: (v: string) => void
  }
}

interface ContextData {
  bucket: BucketContext
  message: MessageContext
  meta: MetaContext
  name: NameContext
  workflow: WorkflowContext
}

function useMessage(): MessageContext {
  const [value, setValue] = React.useState('')
  return React.useMemo(
    () => ({
      state: {
        value,
      },
      actions: {
        onChange: setValue,
      },
    }),
    [value],
  )
}

function useName(src: Src, workflow: WorkflowContext): NameContext {
  const [value, setValue] = React.useState(src.packageHandle?.name || '')
  const state = React.useMemo(
    () => (workflow.state === L ? L : { value }),
    [value, workflow.state],
  )
  return React.useMemo(
    () => ({
      state,
      actions: {
        onChange: setValue,
      },
    }),
    [state],
  )
}

function useMeta(workflow: WorkflowContext, manifest?: Manifest | typeof L): MetaContext {
  const state = React.useMemo(() => {
    if (manifest === L || workflow.state === L) return L
    return {
      value: manifest?.meta,
      schema: workflow.state.value?.schema,
    }
  }, [workflow.state, manifest])
  return React.useMemo(
    () => ({
      state,
    }),
    [state],
  )
}

const Ctx = React.createContext<ContextData | null>(null)

export function useContext(): ContextData {
  const data = React.useContext(Ctx)
  if (!data) throw new Error('Set provider')
  return data
}

interface ProviderProps {
  bucket: string
  name: string
  hashOrTag: string
  hash?: string
  path: string
  mode?: string
  resolvedFrom?: string
  size?: number
  children: React.ReactNode
}

export default function Provider({
  bucket: srcBucket,
  name: srcName,
  hashOrTag,
  path,
  children,
}: ProviderProps) {
  const src = useSource(srcBucket, srcName, hashOrTag, path)
  const bucket = useBucket(src)

  const manifest = useManifest(src)
  const workflow = useWorkflow(bucket.state?.value, manifest)
  const name = useName(src, workflow)
  const message = useMessage()
  const meta = useMeta(workflow, manifest)
  const v = React.useMemo(
    () => ({
      bucket,
      message,
      name,
      workflow,
      meta,
    }),
    [bucket, message, meta, name, workflow],
  )
  return <Ctx.Provider value={v}>{children}</Ctx.Provider>
}
