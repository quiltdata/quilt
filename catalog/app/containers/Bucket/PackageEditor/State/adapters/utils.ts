import type { FileWithPath } from 'react-dropzone'

import { Status } from 'components/FileManager/FileRow'
import type { TreeEntry } from 'components/FileManager/FileTree'
import type * as Model from 'model'
import * as s3paths from 'utils/s3paths'

export function calcChildren(
  entry: Model.PackageEntry | Model.S3File | FileWithPath,
  tailParts: string[],
  children: TreeEntry[] = [],
): TreeEntry[] {
  const [name, ...tail] = tailParts
  const found = children.find((child) => child.id === name)
  if (found) {
    if (tail.length) {
      found.children = calcChildren(entry, tail, found.children)
    }
    return children
  }

  return children.concat({
    id: name,
    name: tail.length ? s3paths.ensureSlash(name) : name,
    size: tail.length ? 0 : entry.size,
    status: Status.Unchanged,
    children: tail.length ? calcChildren(entry, tail, []) : undefined,
  })
}

export function sortEntries(entries: TreeEntry[]): TreeEntry[] {
  return [...entries].sort((a, b) => {
    if (a.children && !b.children) return -1
    if (!a.children && b.children) return 1
    return a.name.localeCompare(b.name)
  })
}
