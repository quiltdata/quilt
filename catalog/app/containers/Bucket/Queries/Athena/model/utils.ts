export const Loading = Symbol('loading')

export type Maybe<T> = T | null

// `T` is the value
// `undefined` is no data. It is not initialized
// `Loading` is loading
// `Error` is error
export type Data<T> = T | undefined | typeof Loading | Error

export interface DataController<T> {
  data: Data<T>
  loadMore: () => void
}

export function wrapData<T>(data: Data<T>, setPrev: (d: T) => void): DataController<T> {
  return {
    data,
    loadMore: () => hasData(data) && setPrev(data),
  }
}

export interface List<T> {
  list: T[]
  next?: string
}

// `T` is the value
// `null` is no value, explicitly set by user
// `undefined` is no value. It is not initialized
// `Loading` is loading
// `Error` is error
export type Value<T> = Maybe<Data<T>>

export interface ValueController<T> {
  value: Value<T>
  setValue: (v: T | null) => void
}

export function wrapValue<T>(
  value: Value<T>,
  setValue: (d: T | null) => void,
): ValueController<T> {
  return {
    value,
    setValue,
  }
}

/** Data is loaded, or value is set to actual value */
export function hasData<T>(value: Value<T>): value is T {
  if (
    value === undefined ||
    value === Loading ||
    value instanceof Error ||
    value === null
  ) {
    return false
  }
  return true
}

/** No value yet: value or data was just initialized */
export function isNone<T>(value: Value<T>): value is undefined {
  return value === undefined
}

/** Data is loading, or value is waiting for data */
export function isLoading<T>(value: Value<T>): value is typeof Loading {
  return value === Loading
}

export function isError<T>(value: Value<T>): value is Error {
  return value instanceof Error
}

/** Value is selected with some or no value, or resolved with error, or data is loaded (successfully or not) */
export function isReady<T>(value: Value<T>): value is T | null | Error {
  if (value === undefined || value === Loading) {
    return false
  }
  return true
}

/** Value is selected with some or no value, or data is loaded successfully */
export function hasValue<T>(value: Value<T>): value is T | null {
  if (value === undefined || value === Loading || value instanceof Error) {
    return false
  }
  return true
}

/** User explicitly set no value */
export function isNoneSelected<T>(value: Value<T>): value is null {
  return value === null
}
