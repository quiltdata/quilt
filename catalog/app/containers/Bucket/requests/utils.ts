import * as R from 'ramda'

export const decodeS3Key = R.pipe(R.replace(/\+/g, ' '), decodeURIComponent)
