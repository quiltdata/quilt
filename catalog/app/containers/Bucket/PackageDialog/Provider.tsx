import * as React from 'react'
import * as RRDom from 'react-router-dom'

import * as workflows from 'utils/workflows'

import {
  PackageCreationDialogUIOptions,
  PackageCreationDialog,
} from '../PackageDialog/PackageCreationForm'

export function CreatePackageLink(props: Omit<RRDom.LinkProps, 'to'>) {
  const location = RRDom.useLocation()
  const search = React.useMemo(() => {
    const searchParams = new URLSearchParams(location.search)
    searchParams.set('createPackage', 'true')
    return searchParams.toString()
  }, [location.search])
  return <RRDom.Link {...props} to={(loc) => ({ ...loc, search: `?${search}` })} />
}

const Ctx = React.createContext<{
  open: () => void
  setDst: (s: workflows.Successor) => void
} | null>(null)

interface ProviderProps {
  // Use it if you need to add CreatePackage dialog for specific places
  // ex. ?createPackage=bookmarks
  // and ignore global ?cretePackage=true
  id?: 'bookmarks'
  bucket: string
  name?: string
  hashOrTag?: string
  s3Path?: string
  ui: PackageCreationDialogUIOptions
  delayHashing?: boolean
  disableStateDisplay?: boolean
  children: React.ReactNode
}

export function Provider({
  id,
  bucket,
  name,
  hashOrTag,
  s3Path,
  ui,
  delayHashing,
  disableStateDisplay,
  children,
}: ProviderProps) {
  const history = RRDom.useHistory()
  const location = RRDom.useLocation()
  const searchParams = React.useMemo(
    () => new URLSearchParams(location.search),
    [location.search],
  )
  const open = React.useCallback(() => {
    searchParams.set('createPackage', id || 'true')
    history.push({
      search: searchParams.toString(),
    })
  }, [id, history, searchParams])
  const [successor, setSuccessor] = React.useState({
    slug: bucket,
  } as workflows.Successor)
  const packageHandle = React.useMemo(() => {
    if (!name || !hashOrTag) return undefined
    return { name, hashOrTag }
  }, [name, hashOrTag])
  const value = React.useMemo(
    () => ({
      open,
      setDst: setSuccessor,
    }),
    [open],
  )
  return (
    <Ctx.Provider value={value}>
      <PackageCreationDialog
        ui={ui}
        id={id}
        src={{ bucket, packageHandle, s3Path }}
        successor={successor}
        onSuccessor={setSuccessor}
        delayHashing={delayHashing}
        disableStateDisplay={disableStateDisplay}
      />
      {children}
    </Ctx.Provider>
  )
}

export const useCreatePackage = () => React.useContext(Ctx)

export const use = useCreatePackage
