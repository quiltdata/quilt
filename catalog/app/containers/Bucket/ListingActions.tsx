import cx from 'classnames'
import * as React from 'react'
import { matchPath, match as Match } from 'react-router-dom'
import * as M from '@material-ui/core'

import * as Bookmarks from 'containers/Bookmarks/Provider'
import * as Model from 'model'
import * as AWS from 'utils/AWS'
import * as BucketPreferences from 'utils/BucketPreferences'
import * as NamedRoutes from 'utils/NamedRoutes'
import * as s3paths from 'utils/s3paths'

import * as FileView from './FileView'

const useButtonStyles = M.makeStyles({
  root: {
    padding: '5px',
  },
})

interface ButtonProps extends M.IconButtonProps {
  download?: boolean
  href?: string
  icon: string
}

function Button({ icon, className, ...props }: ButtonProps) {
  const classes = useButtonStyles()
  return (
    <M.IconButton className={cx(classes.root, className)} {...props}>
      <M.Icon>{icon}</M.Icon>
    </M.IconButton>
  )
}

interface BucketButtonProps {
  className: string
  location: Model.S3.S3ObjectLocation
}

function Bookmark({ className, location }: BucketButtonProps) {
  const bookmarks = Bookmarks.use()
  if (!bookmarks) return null
  const isBookmarked = location ? bookmarks.isBookmarked('main', location) : false
  const toggleBookmark = () => location && bookmarks.toggle('main', location)
  return (
    <Button
      icon={isBookmarked ? 'turned_in' : 'turned_in_not'}
      className={className}
      title="Bookmark"
      onClick={toggleBookmark}
    />
  )
}

function BucketDirectory({ className, location: { bucket, key } }: BucketButtonProps) {
  return (
    <FileView.ZipDownloadForm suffix={`dir/${bucket}/${key}`} className={className}>
      <Button icon="arrow_downward" title="Download" type="submit" />
    </FileView.ZipDownloadForm>
  )
}

function BucketFile({ className, location }: BucketButtonProps) {
  const url = AWS.Signer.useDownloadUrl(location)
  return (
    <Button
      href={url}
      className={className}
      title="Download"
      download
      icon="arrow_downward"
    />
  )
}

interface PackageDirectoryProps {
  className: string
  handle: {
    bucket: string
    name: string
  }
  revision: string
  path: string
}

function PackageDirectory({
  className,
  handle: { bucket, name },
  revision,
  path,
}: PackageDirectoryProps) {
  return (
    <FileView.ZipDownloadForm
      suffix={`package/${bucket}/${name}/${revision}/${path}`}
      className={className}
    >
      <Button icon="arrow_downward" title="Download" type="submit" />
    </FileView.ZipDownloadForm>
  )
}

interface PackageFileProps {
  className: string
  physicalKey: string
}

function PackageFile({ className, physicalKey }: PackageFileProps) {
  const location = React.useMemo(() => s3paths.parseS3Url(physicalKey), [physicalKey])
  return <BucketFile className={className} location={location} />
}

const useRowActionsStyles = M.makeStyles((t) => ({
  root: {
    height: '100%',
    display: 'flex',
    flexDirection: 'row-reverse',
  },
  button: {
    padding: '5px',
  },
  container: {
    background: `linear-gradient(
      to right,
      transparent 0,
      ${t.palette.action.hover} ${t.spacing(4)}px,
      ${t.palette.action.hover} 100%)`,
    display: 'flex',
    padding: t.spacing(0, 2, 0, 10),
  },
  item: {
    '& + &': {
      marginLeft: t.spacing(1),
    },
  },
  wrapper: {
    '.MuiDataGrid-row:hover &': {
      animation: `$show 150ms ease-out`,
      position: 'absolute',
    },
    background: `linear-gradient(
      to right,
      transparent 0,
      ${t.palette.background.paper} ${t.spacing(4)}px,
      ${t.palette.background.paper} 100%)`,
  },
  '@keyframes show': {
    '0%': {
      opacity: 0.3,
    },
    '100%': {
      opacity: '1',
    },
  },
}))

interface BucketMatchParams {
  bucket: string
  path: string
}

interface PackageMatchParams {
  bucket: string
  name: string
  hash?: string
  path: string
}

function useMatchedParams(to: string) {
  const { paths } = NamedRoutes.use()
  return React.useMemo(() => {
    const bucketMatchers = [
      {
        path: paths.bucketFile,
        exact: true,
        strict: true,
      },
      {
        path: paths.bucketDir,
        exact: true,
      },
      {
        path: paths.bucketPackageTree,
        exact: true,
      },
    ]
    const match = bucketMatchers.reduce(
      (memo, matcher) => memo ?? matchPath(to, matcher),
      null as Match<BucketMatchParams | PackageMatchParams> | null,
    )
    if (!match) return {}
    switch (match.path) {
      case paths.bucketFile:
      case paths.bucketDir: {
        const { params } = match as Match<BucketMatchParams>
        return {
          location: {
            bucket: params.bucket,
            key: decodeURIComponent(params.path),
          },
        }
      }
      case paths.bucketPackageTree: {
        const { params } = match as Match<PackageMatchParams>
        return {
          handle: {
            bucket: params.bucket,
            name: params.name,
          },
          revision: params.hash || 'latest',
          path: params.path,
        }
      }
      default:
        throw new Error(`Unexpected path '${match.path}'. Should have been never.`)
    }
  }, [paths, to])
}

interface RowActionsProps {
  to: string
  archived?: boolean
  physicalKey?: string
  prefs: BucketPreferences.ActionPreferences
}

export function RowActions({ archived, physicalKey, to, prefs }: RowActionsProps) {
  const classes = useRowActionsStyles()
  const { location, handle, revision, path } = useMatchedParams(to)

  if (archived) return <></>

  if (location) {
    const DownloadButton = s3paths.isDir(location.key) ? BucketDirectory : BucketFile
    return (
      <div className={classes.root}>
        <div className={classes.wrapper}>
          <div className={classes.container}>
            <Bookmark className={classes.item} location={location} />
            {prefs.downloadObject && (
              <DownloadButton className={classes.item} location={location} />
            )}
          </div>
        </div>
      </div>
    )
  }

  if (handle && prefs.downloadPackage) {
    return (
      <div className={classes.root}>
        <div className={classes.wrapper}>
          <div className={classes.container}>
            {physicalKey ? (
              <PackageFile className={classes.item} physicalKey={physicalKey} />
            ) : (
              <PackageDirectory
                className={classes.item}
                handle={handle}
                revision={revision}
                path={path}
              />
            )}
          </div>
        </div>
      </div>
    )
  }

  return <></>
}
