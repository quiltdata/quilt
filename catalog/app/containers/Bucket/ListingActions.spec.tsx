import * as React from 'react'
import renderer from 'react-test-renderer'
import * as NamedRoutes from 'utils/NamedRoutes'
import * as Bookmarks from 'containers/Bookmarks/Provider'
import { bucketFile, bucketDir, bucketPackageTree } from 'constants/routes'

import { RowActions } from './ListingActions'

jest.mock(
  'constants/config',
  jest.fn(() => ({
    noDownload: false,
    s3Proxy: '',
  })),
)

const defaultPrefs = {
  copyPackage: true,
  createPackage: true,
  deleteRevision: true,
  downloadObject: true,
  downloadPackage: true,
  openInDesktop: true,
  revisePackage: true,
  writeFile: true,
}

function TestBucket({ children }: React.PropsWithChildren<{}>) {
  return (
    <Bookmarks.Provider>
      <NamedRoutes.Provider routes={{ bucketFile, bucketDir, bucketPackageTree }}>
        {children}
      </NamedRoutes.Provider>
    </Bookmarks.Provider>
  )
}

describe('components/ListingActions', () => {
  describe('RowActions', () => {
    it('should render nothing if archived', () => {
      const tree = renderer
        .create(
          <TestBucket>
            <RowActions archived to="" prefs={defaultPrefs} />
          </TestBucket>,
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should render nothing if no route', () => {
      const tree = renderer
        .create(
          <TestBucket>
            <RowActions to="" prefs={defaultPrefs} />
          </TestBucket>,
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should render nothing if wrong route', () => {
      const tree = renderer
        .create(
          <TestBucket>
            <RowActions to="/b/bucketA/BRANCH/fileB" prefs={defaultPrefs} />
          </TestBucket>,
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should render Bucket directory', () => {
      jest.mock('react-redux')
      const tree = renderer
        .create(
          <TestBucket>
            <RowActions to="/b/bucketA/tree/dirB/" prefs={defaultPrefs} />
          </TestBucket>,
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should render Bucket file', () => {
      jest.mock('utils/AWS')
      const tree = renderer
        .create(
          <TestBucket>
            <RowActions to="/b/bucketA/tree/fileB" prefs={defaultPrefs} />
          </TestBucket>,
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should render Package directory', () => {
      const tree = renderer
        .create(
          <TestBucket>
            <RowActions
              to="/b/bucketA/packages/namespaceB/nameC/tree/latest/dirD/"
              prefs={defaultPrefs}
            />
          </TestBucket>,
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should render Package file', () => {
      const tree = renderer
        .create(
          <TestBucket>
            <RowActions
              to="/b/bucketA/packages/namespaceB/nameC/tree/latest/fileD"
              physicalKey="s3://bucketA/pathB/fileB"
              prefs={defaultPrefs}
            />
          </TestBucket>,
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should render Bucket file without download button', () => {
      jest.mock('utils/AWS')
      const tree = renderer
        .create(
          <TestBucket>
            <RowActions
              to="/b/bucketA/tree/fileB"
              prefs={{ ...defaultPrefs, downloadObject: false }}
            />
          </TestBucket>,
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should render Package directory without download button', () => {
      const tree = renderer
        .create(
          <TestBucket>
            <RowActions
              to="/b/bucketA/packages/namespaceB/nameC/tree/latest/dirD/"
              prefs={{ ...defaultPrefs, downloadPackage: false }}
            />
          </TestBucket>,
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })

    it('should render Package file without download button', () => {
      const tree = renderer
        .create(
          <TestBucket>
            <RowActions
              to="/b/bucketA/packages/namespaceB/nameC/tree/latest/fileD"
              physicalKey="s3://bucketA/pathB/fileB"
              prefs={{ ...defaultPrefs, downloadPackage: false }}
            />
          </TestBucket>,
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
