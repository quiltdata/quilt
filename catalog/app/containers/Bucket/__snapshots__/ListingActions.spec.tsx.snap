// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/ListingActions RowActions should render Bucket directory 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="makeStyles-wrapper-5"
  >
    <div
      className="makeStyles-container-3"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root makeStyles-root-7 makeStyles-item-4"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Bookmark"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root"
          >
            turned_in_not
          </span>
        </span>
      </button>
      <form
        action="/zip/dir/bucketA/dirB/"
        className="makeStyles-item-4"
        method="POST"
        style={
          {
            "flexShrink": 0,
          }
        }
      >
        <input
          name="token"
          type="hidden"
          value="ABC"
        />
        <button
          className="MuiButtonBase-root MuiIconButton-root makeStyles-root-7"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          title="Download"
          type="submit"
        >
          <span
            className="MuiIconButton-label"
          >
            <span
              aria-hidden={true}
              className="material-icons MuiIcon-root"
            >
              arrow_downward
            </span>
          </span>
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`components/ListingActions RowActions should render Bucket file 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="makeStyles-wrapper-5"
  >
    <div
      className="makeStyles-container-3"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root makeStyles-root-7 makeStyles-item-4"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Bookmark"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root"
          >
            turned_in_not
          </span>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
      <a
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root makeStyles-root-7 makeStyles-item-4"
        download={true}
        href="s3://bucketA/fileB"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Download"
      >
        <span
          className="MuiIconButton-label"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root"
          >
            arrow_downward
          </span>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </a>
    </div>
  </div>
</div>
`;

exports[`components/ListingActions RowActions should render Bucket file without download button 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="makeStyles-wrapper-5"
  >
    <div
      className="makeStyles-container-3"
    >
      <button
        className="MuiButtonBase-root MuiIconButton-root makeStyles-root-7 makeStyles-item-4"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Bookmark"
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root"
          >
            turned_in_not
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`components/ListingActions RowActions should render Package directory 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="makeStyles-wrapper-5"
  >
    <div
      className="makeStyles-container-3"
    >
      <form
        action="/zip/package/bucketA/namespaceB/nameC/latest/dirD/"
        className="makeStyles-item-4"
        method="POST"
        style={
          {
            "flexShrink": 0,
          }
        }
      >
        <input
          name="token"
          type="hidden"
          value="ABC"
        />
        <button
          className="MuiButtonBase-root MuiIconButton-root makeStyles-root-7"
          disabled={false}
          onBlur={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          title="Download"
          type="submit"
        >
          <span
            className="MuiIconButton-label"
          >
            <span
              aria-hidden={true}
              className="material-icons MuiIcon-root"
            >
              arrow_downward
            </span>
          </span>
        </button>
      </form>
    </div>
  </div>
</div>
`;

exports[`components/ListingActions RowActions should render Package directory without download button 1`] = `null`;

exports[`components/ListingActions RowActions should render Package file 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="makeStyles-wrapper-5"
  >
    <div
      className="makeStyles-container-3"
    >
      <a
        aria-disabled={false}
        className="MuiButtonBase-root MuiIconButton-root makeStyles-root-7 makeStyles-item-4"
        download={true}
        href="s3://bucketa/pathB/fileB"
        onBlur={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Download"
      >
        <span
          className="MuiIconButton-label"
        >
          <span
            aria-hidden={true}
            className="material-icons MuiIcon-root"
          >
            arrow_downward
          </span>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </a>
    </div>
  </div>
</div>
`;

exports[`components/ListingActions RowActions should render Package file without download button 1`] = `null`;

exports[`components/ListingActions RowActions should render nothing if archived 1`] = `null`;

exports[`components/ListingActions RowActions should render nothing if no route 1`] = `null`;

exports[`components/ListingActions RowActions should render nothing if wrong route 1`] = `null`;
